synapse/handlers/user_directory.py

```diff
@@ -436,14 +400,15 @@ def _handle_room_publicity_change(self, room_id, prev_event_id, event_id, typ):
             # ignore the change
             return
 
-        if change:
-            users_with_profile = yield self.state.get_current_user_in_room(room_id)
-            for user_id, profile in iteritems(users_with_profile):
-                yield self._handle_new_user(room_id, user_id, profile)
-        else:
-            users = yield self.store.get_users_in_public_due_to_room(room_id)
-            for user_id in users:
-                yield self._handle_remove_user(room_id, user_id)
+        users_with_profile = yield self.state.get_current_user_in_room(room_id)
+
+        # Remove every user from the sharing tables for that room.
+        for user_id in iterkeys(users_with_profile):
+            yield self.store.remove_user_who_share_room(user_id, room_id)
+
+        # Then, re-add them to the tables.
+        for user_id, profile in iteritems(users_with_profile):
+            yield self._handle_new_user(room_id, user_id, profile)
```

Comment:
it feels like we're (still) repeating ourselves here.

Suppose there are 3 local users in the room: A, B, C.

First we call _handle_new_user with A. That will add entries: (A, A), (A, B), (A, C), (B, A), (C, A).
Then we call _handle_new_user with B. That will add entries: (B, B), (B, A), (B, C), (A, B), (C, B).
Finally we call _handle_new_user with C. That will add entries: (C, C), (C, A), (C, B), (A, C), (B, C).

So most of those entries are being added twice.

(happy if you want to say that's an existing problem, to be ignored for now, but since we're rewriting this it seems a reasonable time to consider it, or at least add a comment)

Reply:
Without the adding yourself, that drops one of those, but yes, it does repeat that, but that's an existing problem.

Reply:
> Without the adding yourself, that drops one of those

Ironically the "adding yourself" was the only entry which *wasn't* being redone.

> yes, it does repeat that, but that's an existing problem.

Fair enough. I'd still be happier if you could add a TODO or something so that I know I'm not going mad next time I look at this code.